File: Offer/Offer04.java
public class Offer04 {

//    public boolean findNumberIn2DArray(int[][] matrix, int target) {
//        boolean res = false;
//
//        if (matrix == null || matrix.length == 0 || matrix[0].length == 0){
//            return res;
//        }
//
//
//        for(int i = 0; i < matrix.length; i++){
//            for (int j = 0; j < matrix[0].length; j++){
//                if (matrix[i][j] == target){
//                        res = true;
//                }
//            }
//        }
//        return res;
//    }

    public boolean findNumberIn2DArray(int[][] matrix, int target){
        boolean res = false;
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0){
            return res;
        }

        int i = matrix.length-1,j = 0;

        while (i >= 0 && j < matrix[0].length){
            if (matrix[i][j] > target){
                i--;
            }else if (matrix[i][j] < target) j++;
            else{
                res = true;
                return res;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Offer04 offer04 = new Offer04();
        int arr[][] = {{-1,3}};
        System.out.println(offer04.findNumberIn2DArray(arr,3));
    }
}

File: Offer/C++/offer04.cc
#include<iostream>
#include<vector>
#include<queue>
using namespace std;

class Solution {
public:
    // BFS做法
    bool findNumberIn2DArray(vector<vector<int>>& matrix, int target) {
       if(matrix.size() == 0 || matrix[0].size() == 0){
           return false;
       }

       if(matrix[0][0] == target){
           return true;
       }
       
       int nums = matrix[0][0];
       queue<pair<int,int>> q;
       int m = matrix.size();
       int n = matrix[0].size();
       vector<int> dir = {0,1,0,-1,0};
       q.push(make_pair(0,0));
       matrix[0][0] = nums; //visited

       while(!q.empty()){

           auto coord = q.front();
           q.pop();

           for(int i = 0; i < 4; i++){

               int row =  coord.first+dir[i];
               int col = coord.second+dir[i+1];

               if(row >= 0 && row < m && col >=0 && col < n){
                   if(matrix[row][col] == target){
                       return true;
                   }
                   if(matrix[row][col] != nums){
                       q.push(make_pair(row,col));
                       matrix[row][col] = nums;
                   }
               }
           }
       }
       return false;
    }
};

int main()
{
    return 0;
}
